### ISSUE TITLE
Create function Context.get(socket, ...) to access the context

### ISSUE THREAD

**Issue by dvic** [surface-ui/surface#260]
Is there a public API for accessing the context inside the `update` callback of a component?

**Comment by ewildgoose** [comment]
Do you mean that, in the case of a sub component you want to see the context provided by the upper component?

I think the insight on components is that as you descend, each level only has access to anything *explicitly* passed down by the level above. ie there is no concept of a global variable that's available across all components. 

Now, each component gets it's "assigns" trimmed as it gets called by a tier above. However, there is both explicit access to the assigns (through "prop" and "data"), but also an optional behaviour through "contexts". These seem to set sticky params in the assigns, which can then be (optionally) fished out and accessed in nested components

So I *think* in answer to your question, what you can see in the assigns is what you have access to. Context set by higher levels would be visible (I think). However, I'm unclear if you can alter them safely in a call back? But I guess reading them is acceptable?

**Comment by dvic** [comment]
> Do you mean that, in the case of a sub component you want to see the context provided by the upper component?

Yes, you can do this of course in the template, but I was wondering if there is an API / plans to read it from the socket (i.e., assigns) in the `update/2` callback of a component.

@msaraiva I could make a PR that adds an API to read the context from the socket assigns, but I'm not sure if this is something you'd like to have in the lib (maybe I'm missing the intended usage of `Context`)

**Comment by msaraiva** [comment]
Hi @dvic!

Sorry for the late response. I think we could have a `Context.get(socket, ...)`. As far as I know, @olivermt is already using this approach in `surface_bootstrap`.  @lnr0626 do you think it's time to add this, even though we'll not be able to provide a `Context.put(...)`?

**Comment by lnr0626** [comment]
Yeah, that makes sense - not having the put api is reasonable, but a get would be useful

**Comment by dvic** [comment]
Cool, I'll try to cook up a PR for this.

**Comment by olivermt** [comment]
@dvic Hi! Marlus  sent me here to comment and start a discussion :)

Question 1 is why is there a conclusion about not having put in that issue? I don't get it.

Then to follow up from that I feel that we should take this one step further than just adding some functions, especially I think we should have a proper discussion on what we want Context to be and make some decisions before we expose a public API that will lock in what is possible.

I have some thoughts and deliberations and I feel that we need to cover them before we revisit your PR.

- The way context currently work is already not very friendly to re-rendering decisions etc so anything we officially do around Context should probably be related to making the usage more declarative and probably exposed through a data prop or similar
- The `Process` seems to have become the stable vehicle for liveviews and we can probably expect that a liveivew is always one to two processes (the dead first, then the live on ws connect or live redirect).
- Due to process being a stable vehicle, a context could possibly be rooted in the pdict instead of a macroed "pass it down the pipe" approach. This gives true prop drilling. 
- To support using the pdict and also support the rendering engine actually understanding whats up, we need to have some strong guard rails around how you use a context value and how to let the context provider know that a value was actually used (so we can signal re-rendering down the tree)

From all this, I propose that we expand the data prop with a `provider` configuration. The reason I am saying provider is that it is also extensible to more things later. If we now expand with `provider: :context` we can later do something like `provider: {:pub_sub, "some-topic"}` or `provider: &my_dataloading_fun/2` where `socket` is passed as first arg.

Another reason to do this is to lift the Context plumbing out of the "unofficial" assigns and into a proper prop. If Context.put puts a context assign in the pdict of the liveview, any child component be it stateful or stateless can Context.get and grab it out of the pdict. 

Now this is where the framework aspect comes into play and why I want a declarative prop. This means that any component that uses a dataprop with that provider can be tracked by the compiler so that we can enforce re-renders down the tree of modules that has used a given context provided data prop.

This last part is where I need people with bigger macro / framework brains than I have to give their five cent. I think we *might* be skirting close to some framework magic that everyone tried to escape in Rails by going to the more explicit and declarative Elixir, but I still think this is cleaner than trying to sidestep the whole assigns internal thing going on in LiveView with some spaghetti context internal assign.

Some open questions on my end is:
1. Can we given a module name (like `Foo.BarComponent`) that has used a context prop hook it in any way to either re-render or kick off its update/2 cycle? This is probably very achievable with stateful components just delegating down to send_update, but I am not sure how to solve this with stateless components, as I am not quite sure what their lifecycle is at all internally in Surface/LiveView.
2. Is this really a better solution than passing `__context__` everywhere? The questions we need to solve around being declarative and making sure that change tracking and needless re-rendering (as in we need to expose an actual `@variable` so that the leex engine can state diff) are the same regardless of the vehicle we use to push context down. Forms rely on context being unique in each Context branch, so we need to solve that somehow if we go the liveview pdict route, so maybe nested macro assign context is better either way?

Looking forward to discussion :)

**Comment by harmon25** [comment]
We should pay attention to this live_view issue, as it plays into this context discussion a bit for non change tracked context: https://github.com/phoenixframework/phoenix_live_view/issues/1386

As much as I like the pdict idea, ( have tried it myself). We should avoid it as part of the core - as its a path that will be hard to back out of. If storing data in the pdict is exposed via some opt in plugin system or context provider idea as proposed it would be OK

**Comment by olivermt** [comment]
I have also decided that I would like to backtrack the pdict idea.

80% of my post still stands nicely in terms of exposing declarative props etc if we base it on macro based assigns propagation of a 'hidden' assign though.

**Comment by lnr0626** [comment]
> why is there a conclusion about not having put in that issue?

the tldr is because of how context is handled currently providing a put is non trivial - while it's potentially feasible, the ways that come to mind would be even less friendly to change tracking than what we currently allow :-(

> We should pay attention to this live_view issue ...

That's an interesting concept - I guess the idea would be to use whatever solution they have to store the surface data (that shouldn't be change-tracked) like `__context__`, and create assigns when specific pieces of it are accessed that would be changed-tracked correctly

We may still have some issues around slots - iirc implementing context so that it can be propagated to slot content does not play nicely with change-tracking; though it's been a long time since I've been in that area of the code and it's feasible live view has improved since then

> given a module name (like Foo.BarComponent) that has used a context prop, can we hook it in any way to either re-render or kick off its update/2 cycle? ...

My understanding is that stateless components are re-rendered based on when the stateful thing (live view/stateful live component) that they are embedded in is re-rendered. I also believe that in the stateful component case, we need the component id as well as the module name; though I don't know for sure and I'm happy to be proven wrong :-)

> Another reason to do this is to lift the Context plumbing out of the "unofficial" assigns and into a proper prop

A long while ago, we had something close to this. Instead of a `data` prop, there was a `context` macro and an `init_context` method that had to return the values for context props. We decided to re-implement it as a component with some compiler support. One of the interesting capabilities with our current implementation is the ability to have multiple different context trees within the same component - though I'm unsure if that's actually used anywhere

> From all this, I propose that we expand the data prop with a provider configuration

This is an intriguing idea, and allows some interesting ideas like you hinted at with other providers (and potentially allows library authors/users to create their own providers). I think we'd quickly run into the question of how to handle initialization order (currently `prop` and `data` ordering doesn't have any meaning to surface itself, but once we allow automatically initializing values it would potentially matter).

> The questions we need to solve around being declarative and making sure that change tracking and needless re-rendering (as in we need to expose an actual @variable so that the leex engine can state diff)

It would be nice to also address the issue underlying #267 as well - it's not immediately obvious that rendering a component inline and calling a function to render the same component have different results if the context is being used; and even when you know that, there's not really a satisfactory reason (at least to me) :-(

**Comment by dvic** [comment]
@olivermt Thanks for starting the discussion! 

I'm not so familiar with the code base but after reading the responses above I do have a better intuition of the current limitations and what the possible directions are, so thanks for that!

If I would have to summarize the above macro-based approach, would that be something like:

```elixir
defmodule MyComponent do
  context MyContext, :foo

  @impl
  def update(assigns, socket) do
    socket = assign(socket, assigns)
    
    foo = get_context(MyContext, :foo)

    {:ok, socket}
  end
end
```

where `get_context` is a macro that generates code that fails at compile-time if the `context` attribute is missing for that key and if it's not missing generates `Map.fetch!(socket.assigns.__context_, {MyContext, :foo})` (or something like that).

**Comment by msaraiva** [comment]
Next version of LV will ship with some new features that might allow us to implement a better solution for contexts. If that turns out to be true, we might rethink the way we handle context assigns and maybe come up with a better API that maybe doesn't need this function.

I'm closing this for now. We can re-open it later if we think it can still be useful.

**Comment by maennchen** [comment]
When this feature is thought through in the future, I'd like to show one more use-case that I think should be considered:

https://github.com/jshmrtn/hygeia/blob/f5c578981251ec9245a98b88ee972e4d9ed57e8a/apps/hygeia_web/lib/hygeia_web/live_view.ex#L74

Our own LiveView module extends on `Phoenix.LiveLive` & `Surface.LiveView`. Its job is to inject some global values into the context. Those are:
* `auth` (user etc.)
* `logged_in` (convenience instead of checking `is_nil(auth)`
* `browser_features` (relayed modernizr flags to detect support of browser features like date inputs)
* `ip_address` (request IP)
* `uri` (request URI; used to show active link in navigations)
* `params` (path params for current URL; used to show active link in navigations)

Since we didn't want to repeat a `<Context put={...}>` in every component, we inject those directly into the `assigns` at `mount` / `handle_params`.

I'm not sure if this is the nicest solution, but if a feature is planned to replace contexts, I'd like to consider a use case like this as well.

**Comment by maennchen** [comment]
PS @msaraiva: What new LiveView feature are you referring to? `live_session`?

**Comment by Malian** [comment]
@maennchen From what I have seen on Twitter, it seems that Phoenix Live View will introduce "hooks" or "callbacks" to, hopefully, perform some manipulations before and after mount and update functions.

I implemented in #336 a first try of a plugin system that is pretty close to what you have done with Hygeia. We also closed this PR waiting for more info from LiveView.

**Comment by maennchen** [comment]
@Malian I switched to using the new Phoenix LiveVIew Hooks in Hygeia. I however still see the need for a `Contact.put` function.

For reference:

* My `Context.put`: https://github.com/jshmrtn/hygeia/blob/e8c6c5bd8774dbe61fa2397e1386d06f9f233530/apps/hygeia_web/lib/hygeia_web.ex#L205
* Usage Example: https://github.com/jshmrtn/hygeia/blob/e8c6c5bd8774dbe61fa2397e1386d06f9f233530/apps/hygeia_web/lib/hygeia_web/live/init/context.ex#L38

Would you be interested in a PR that exposes this directly on the surface `Context` module?

**Comment by dvic** [comment]
> Would you be interested in a PR that exposes this directly on the surface `Context` module?

@maennchen I like it! However, I wonder: does it also work when you have a component that *only* accesses the context from mount/update? Does surface currently always pass the `__context__` assign? I believe I encountered this issue previously, where the `__context__` assign would only be passed when I used `<Context>` in the component template.

**Comment by maennchen** [comment]
@dvic I've used this everywhere in our application. Both through the component and also reading from the socket. I never encountered any problems with it so far.

**Comment by dvic** [comment]
> Both through the component and also reading from the socket. I never encountered any problems with it so far.

@maennchen Yes when you have *both* it works, but do you also have an example of a component that *only* reads it using `context_get` (i.e., from the socket) and does not have `<Context>` in the template?

I expect this not to work for that component because it does not have the `__context__` assign passed, because Surface optimizes this away (unless the implementation changes since the last time I checked).

**Comment by maennchen** [comment]
@dvic This example should definitely be refactored, but it currently exclusively reads directly from the context:

https://github.com/jshmrtn/hygeia/blob/1481ccedf8b00889113eaf8fd9227026551219ad/apps/hygeia_web/lib/hygeia_web/live/record_view.ex#L57

**Comment by dvic** [comment]
@maennchen I just checked with an [example repo](https://github.com/qdentity/surface_playground/tree/dvic-surface-context) and in my case it doesn't work. If you remove the [`<Context>` line](https://github.com/qdentity/surface_playground/blob/dvic-surface-context/lib/surface_playground_web/live/example.ex#L61), the `__context__` assign is empty inside the `update/2` callback, just as I expected. I'm not sure why the context is working in your `RecordView` component though.

**Comment by maennchen** [comment]
@dvic I think the update is only called with changes. Could it be that you didn't merge the current assigns and the updated ones to check?

(I have not checked how context actually works, I just assumed it's working like a normal `assign`...)

**Comment by dvic** [comment]
@maennchen As far as I know the `update/2` callback is also called on first render (see LiveComponent lifecycle [docs](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveComponent.html#module-life-cycle)) and I do assign the assigns, on [line 50](https://github.com/qdentity/surface_playground/blob/dvic-surface-context/lib/surface_playground_web/live/example.ex#L50). I also see the context being present in the `assigns` argument in `update/2` just as long as I have the `<Context get...>` line in the template 🙂.  

This is something that has bitten me previously, that's why I [proposed](https://github.com/surface-ui/surface/issues/260#issuecomment-825559351) a declarative macro as a second option to "signal" to the surface compiler that a component is dependent on some context. But I'm not sure what the current plans are for the context feature of surface.

**Comment by dvic** [comment]
By the way: it kind of makes sense that Surface does this, otherwise LiveView would have to re-render the whole component tree anytime something changes in a context (because all components would be dependent on the whole context). Instead, surface only passes the necessary context keys to each component, and it does this currently based on the `<Context get={}>` presence in the templates.

**Comment by maennchen** [comment]
@dvic I see. Then I somehow got lucky here...

**Comment by msaraiva** [comment]
> does it also work when you have a component that _only_ accesses the context from mount/update? Does surface currently always pass the `__context__` assign? I believe I encountered this issue previously, where the `__context__` assign would only be passed when I used `<Context>` in the component template.

@dvic you're correct. In order to optimize the data sent, Surface only passes the `__context__` assign when it detects the context will be used.  In order to do that, Surface follows a few rules, one of them is to check the `:gets_context?` and `:changes_context?` module attributes. That means is we want to add `Context.put` and `Context.get`, they'll have to be macros so we would have a chance to set those values at compile-time. Like we do here: https://github.com/surface-ui/surface/blob/c33bb3abbb92315585ccde7a58c342d548c01e36/lib/surface/compiler/eex_engine.ex#L965-L971

One of the reasons I want to postpone implementing this feature is because as we're currently implementing slots as a core feature in LV, there's a big chance we'll be able to fix the main issue, which is using variables in slots. If that happens, we'll able to remove all that optimiztion logic from our side, allowing us to rethink contexts and maybe come up with a better API.

BTW, I can see you found an issue with that logic and reported https://github.com/surface-ui/surface/issues/493. I still have to check that.

**Comment by dvic** [comment]
Thanks for the explanation and pointers @msaraiva! I guess it's best to wait then for the slot feature to land in LV, curious to see what it will bring us 🙂

> BTW, I can see you found an issue with that logic and reported #493. I still have to check that.

Not sure if it's exactly related to this, but it is definitely a compiler bug as I just encountered the issue also on the [reproduction repo](https://github.com/qdentity/surface_playground/tree/dvic-surface-context) (now the repo is on `0.5.x` and everything is fine).

By the way: do you happen to know how is it possible that Mix is okay with me having LiveView 0.16 while Surface is on `0.5.x`, which requires LiveView `0.15`? I was expecting to have to do `override: true` for the LiveView dependency.

**Comment by msaraiva** [comment]
@dvic Surface is also compatible with LV `v0.16`. In fact, it's even recommended to use it as there's a bug in LV's diff tracker regarding using nested slots that has only been fixed in `v0.16`.

**Comment by dvic** [comment]
> @dvic Surface is also compatible with LV `v0.16`. In fact, it's even recommended to use it as there's a bug in LV's diff tracker regarding using nested slots that has only been fixed in `v0.16`.

Didn't know that, thanks!

**Comment by kingdomcoding** [comment]
Now that slots are implemented in LiveView, is this feature now in view?

**Comment by msaraiva** [comment]
Hi @kingdomcoding!

Yes, the context API is being currently reviewed and should include a `Context.get(...)` function in future versions.

**Comment by dvic** [comment]
For those interested: this has now been implemented in #641
